common/prove/witness_helper.go

```diff
@@ -597,3 +605,157 @@ func (w *WitnessHelper) constructSimpleWitnessInfo(oTx *tx.Tx) (
 	}
 	return accountKeys, accountWitnessInfo, nftWitnessInfo, nil
 }
+
+func (w *WitnessHelper) ConstructGasWitness(block *block.Block) (cryptoGas *GasWitness, stateRoot []byte, err error) {
+	var gas *circuit.Gas
+
+	needGas := false
+	gasBefore := make(map[int64]*big.Int)
+	gasChanges := make(map[int64]*big.Int)
+	for _, assetId := range types.GasAssets {
+		gasChanges[assetId] = types.ZeroBigInt
+	}
+	for _, tx := range block.Txs {
+		if types.IsL2Tx(tx.TxType) {
+			needGas = true
+			for _, txDetail := range tx.TxDetails {
+				if txDetail.IsGas {
+					assetBefore, err := types.ParseAccountAsset(txDetail.Balance)
+					if err != nil {
+						return nil, nil, err
+					}
+					if _, ok := gasBefore[assetBefore.AssetId]; !ok {
+						gasBefore[assetBefore.AssetId] = assetBefore.Balance
+					}
+					assetDelta, err := types.ParseAccountAsset(txDetail.BalanceDelta)
+					if err != nil {
+						return nil, nil, err
+					}
+					gasChanges[assetDelta.AssetId] = ffmath.Add(gasChanges[assetDelta.AssetId], assetDelta.Balance)
```

Comment:
For the witness, we change the gas account's balance at the block level, we should also do this for the committer, otherwise, if the gas account sends a transfer gas asset transaction, it may succeed in the committer, but failed in the witness.

Reply:
fixed